miner/algo_greedy_buckets.go

```diff
@@ -0,0 +1,211 @@
+				usedEntry.Success = false
+				usedSbundles = append(usedSbundles, usedEntry)
+				CheckRetryOrderAndReinsert(order, orders, retryMap, retryLimit)
+				continue
+			}
+
+			log.Trace("Included sbundle", "bundleEGP", sbundle.MevGasPrice.String(), "ethToCoinbase", ethIntToFloat(sbundle.Profit))
+			usedEntry.Success = true
+			usedSbundles = append(usedSbundles, usedEntry)
+		} else {
+			// note: this should never happen because we should not be inserting invalid transaction types into
+			// the orders heap
+			panic("unsupported order type found")
+		}
+	}
+	return usedBundles, usedSbundles
+}
+
+func (b *greedyBucketsBuilder) mergeOrdersIntoEnvDiff(
+	envDiff *environmentDiff, orders *types.TransactionsByPriceAndNonce) ([]types.SimulatedBundle, []types.UsedSBundle,
+) {
+	if orders.Peek() == nil {
+		return nil, nil
+	}
+
+	const retryLimit = 1
+
+	var (
+		baseFee      = envDiff.baseEnvironment.header.BaseFee
+		retryMap     = make(map[*types.TxWithMinerFee]int)
+		gasUsedMap   = make(map[*types.TxWithMinerFee]uint64)
+		usedBundles  []types.SimulatedBundle
+		usedSbundles []types.UsedSBundle
+		transactions []*types.TxWithMinerFee
+		percent      = big.NewFloat(0.9)
+
+		CutoffPriceFromOrder = func(order *types.TxWithMinerFee) *big.Int {
+			floorPrice := new(big.Float).
+				Mul(
+					new(big.Float).SetInt(order.Price()),
+					percent,
+				)
+			round, _ := floorPrice.Int64()
+			return big.NewInt(round)
+		}
+
+		IsOrderInPriceRange = func(order *types.TxWithMinerFee, minPrice *big.Int) bool {
+			return order.Price().Cmp(minPrice) >= 0
+		}
+
+		SortInPlaceByProfit = func(baseFee *big.Int, transactions []*types.TxWithMinerFee, gasUsedMap map[*types.TxWithMinerFee]uint64) {
+			sort.SliceStable(transactions, func(i, j int) bool {
+				return transactions[i].Profit(baseFee, gasUsedMap[transactions[i]]).Cmp(transactions[j].Profit(baseFee, gasUsedMap[transactions[j]])) > 0
+			})
+		}
+	)
+
+	minPrice := CutoffPriceFromOrder(orders.Peek())
+	for {
+		order := orders.Peek()
+		if order == nil {
+			if len(transactions) != 0 {
+				SortInPlaceByProfit(baseFee, transactions, gasUsedMap)
+				bundles, sbundles := b.commit(envDiff, transactions, orders, gasUsedMap, retryMap, retryLimit)
+				usedBundles = append(usedBundles, bundles...)
+				usedSbundles = append(usedSbundles, sbundles...)
+				transactions = nil
+				// re-run since committing transactions may have pushed higher nonce transactions, or previously
+				// failed transactions back into orders heap
+				continue
+			}
+			break
+		}
+
+		if ok := IsOrderInPriceRange(order, minPrice); ok {
+			orders.Pop()
+			transactions = append(transactions, order)
+		} else {
+			if len(transactions) != 0 {
+				SortInPlaceByProfit(baseFee, transactions, gasUsedMap)
+				bundles, sbundles := b.commit(envDiff, transactions, orders, gasUsedMap, retryMap, retryLimit)
+				usedBundles = append(usedBundles, bundles...)
+				usedSbundles = append(usedSbundles, sbundles...)
```

Comment:
do you handle retry correctly? there is an assumption that you will return used bundles where one bundle will be there once

Reply:
we only insert sbundle after reinsert fails such that sbundle is only appended once